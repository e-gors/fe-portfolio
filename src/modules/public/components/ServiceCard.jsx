import { Avatar, Box, Paper, Typography } from "@mui/material";
import React from "react";
import { OutlinedButton } from "../../../components/CustomButtons";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";

function ServiceCard({ service = "Title", descriptions = [], image }) {
  const theme = useSelector((state) => state.theme.theme);

  // State to track if descriptions are expanded or not
  const [isExpanded, setIsExpanded] = React.useState(false);

  // Handle toggle between expand/collapse
  const toggleExpand = () => {
    setIsExpanded(!isExpanded);
  };

  return (
    <Paper
      sx={{
        borderRadius: 2,
        padding: 2,
        boxShadow: theme === "light" ? 4 : 0,
        position: "relative",
        backgroundColor:
            theme === "light" ? "var(--bg_white)" : "var(--bg_black)",
      }}
    >
      <Box
        sx={{
          position: "absolute",
          top: "-15%",
          left: "40%",
          backgroundColor: "white",
          border: "1px solid #10fddd",
          borderRadius: "50%",
          width: 60,
          height: 60,
          boxShadow: 5,
        }}
      >
        <Avatar
          src={image}
          alt={service}
          sx={{
            position: "relative",
            top: "50%",
            left: "50%",
            transform: "translate(-50%, -50%)",
            width: "70%",
            height: "70%",
          }}
        />
      </Box>
      <Box sx={{ my: 2 }}>
        <Typography variant="subtitle2" gutterBottom>
          {service}
        </Typography>
        <Box
          sx={{
            display: "-webkit-box",
            WebkitBoxOrient: "vertical",
            overflow: "hidden",
            textOverflow: "ellipsis",
            WebkitLineClamp: isExpanded ? "none" : 3, // Show full or truncate text
            minHeight: 80,
            height: "auto",
            maxHeight: isExpanded ? "none" : 100, // Control height based on expand state
            transition: "0.3s all ease",
          }}
        >
          {descriptions?.map((description, i) => (
            <Typography
              key={i}
              variant="body2"
              color="text.secondary"
              component="li"
            >
              {description}
            </Typography>
          ))}
        </Box>
      </Box>
      <OutlinedButton theme={theme} variant="outlined" onClick={toggleExpand}>
        {isExpanded ? "Show Less" : "Read More"} {/* Toggle button text */}
      </OutlinedButton>
    </Paper>
  );
}

ServiceCard.propTypes = {
  service: PropTypes.string,
  descriptions: PropTypes.array,
  icon: PropTypes.node,
};

export default ServiceCard;
