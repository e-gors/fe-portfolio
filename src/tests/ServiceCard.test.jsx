// ServiceCard.test.js
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ServiceCard from '../modules/public/components/ServiceCard';

// Mock icon for testing
const mockIcon = '../assets/svg/dropbox.svg';

describe('ServiceCard', () => {
  it('renders the service card with default props', () => {
    render(<ServiceCard icon={mockIcon} />);

    expect(screen.getByAltText('SVG')).toHaveAttribute('src', mockIcon);
    expect(screen.getByText(/Title/i)).toBeInTheDocument();
    expect(screen.getByText(/Descriptions.../i)).toBeInTheDocument();
  });

  it('renders the service card with provided props', () => {
    const title = 'Service Title';
    const descriptions = 'Detailed description of the service.';
    render(<ServiceCard title={title} descriptions={descriptions} icon={mockIcon} />);

    expect(screen.getByAltText('SVG')).toHaveAttribute('src', mockIcon);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(descriptions)).toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<ServiceCard title="Snapshot Test" descriptions="Snapshot description" icon={mockIcon} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
