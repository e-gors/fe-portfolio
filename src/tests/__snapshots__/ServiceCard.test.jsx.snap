// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceCard matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-15m38mt-MuiPaper-root"
  >
    <div
      class="MuiBox-root css-1sw2ahh"
    >
      <img
        alt="SVG"
        height="60"
        src="../assets/svg/dropbox.svg"
        width="60"
      />
    </div>
    <div
      class="MuiBox-root css-1qm1lh"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-1oxen2w-MuiTypography-root"
      >
        Snapshot Test
      </h6>
      <div
        class="MuiBox-root css-wszdhj"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Snapshot description
        </p>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1gopn71-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Read More
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
